provider/openstack/provider.go

```diff
@@ -1173,55 +1158,28 @@ func (e *Environ) StartInstance(args environs.StartInstanceParams) (*environs.St
 		}
 		inst.floatingIP = publicIP
 	}
+
 	return &environs.StartInstanceResult{
 		Instance: inst,
 		Hardware: inst.hardwareCharacteristics(),
 	}, nil
 }
 
-func (e *Environ) startInstanceAvailabilityZones(args environs.StartInstanceParams) ([]string, error) {
+func (e *Environ) startInstanceAvailabilityZone(args environs.StartInstanceParams) (string, error) {
```

Comment:
To deal with zone placement and restrictions like those introduced by volume attachments, I think the provisioner needs to call into the provider to return a list of zones that it should attempt for that machine. The provisioner will then pick only from that list, still based on population size sorting.

Reply:
The provisioner code uses a method in provider/common named AvailabilityZoneAllocations() which returns the availability zones which are "available" and which instances are currently in what AZ.  Currently the providers are doing the verification that the provisioner suggested zone is appropriate for any volume.